doc: add n28950 notes and questions (#743)
bitcoin-core-review-club/website: _posts/2024-02-07-#28950.md

title: "Add `maxfeerate` and `maxburnamount` args to submitpackage"
pr: 28950
authors: [instagibbs]

> components: ["rpc/rest/zmq"]

Reviewer: In the future I'd recommend using the new post [script](https://github.com/bitcoin-core-review-club/website/blob/master/hosting.md#making-a-new-post), as it would scrape the labels and prevent things like #744.

components: ["rpc"]
host: ismaelsadeeq
status: upcoming
commit: bf6fbea9794db5e8d0adce8ff6182574808cea54
---

## Notes

> - The [`submitpackage RPC`](https://bitcoincore.org/en/doc/26.0.0/rpc/rawtransactions/submitpackage/) RPC method takes `package` as a required argument. The package must consist of a child with its parents, and none of the parents may depend on one another. The package will be validated according to consensus and mempool policy rules. If all transactions pass, they will be accepted to mempool.

Reviewer: Why copy the docs here? Just the link is fine.

Author: Yes fixed 👍🏾

> - Package transactions submitted to the `submitpackage` RPC successfully are accepted to the node's mempool, but broadcasted to peers individually because [Package relay](https://bitcoinops.org/en/topics/package-relay/) is currently a work in progress. Successful submission does not mean the transactions will propagate throughout the network whenever the transactions need to be relayed as a package.

Reviewer: ```suggestion
- Package transactions submitted to the `submitpackage` RPC successfully are accepted to the node's mempool, but broadcasted to peers individually because [Package relay](https://bitcoinops.org/en/topics/package-relay/) is currently a work in progress. Successful submission does not mean the transactions will propagate.
```

Reviewer: Also not sure how this is relevant to the PR. I think it's more relevant to mention the CPFP use case (which can lead to overshooting on fees) instead. That might make this text feel less out of place.

Author: Removed

> - `maxfeerate` is an optional parameter for the `sendrawtransaction` and `testmempoolaccept` RPCs. It prevents transactions from being added to mempool and broadcasted to peers if the transaction feerate exceeds `maxfeerate`, avoiding unintentional payment of a high fee rate.

Reviewer: You're using the term in its own definition here. Also, the parameter itself doesn't do the preventing.
```suggestion
- `maxfeerate` is an optional parameter for the `sendrawtransaction` and `testmempoolaccept` RPCs. Transactions exceeding this feerate won't be added to mempool and broadcasted to peers, helping prevent unintentional overpayment.
```

> - `maxburnamount` is an optional parameter for the `sendrawtransaction` RPC. It prevents broadcasting transactions with a data carrier `OP_RETURN` output amount that exceeds `maxburnamount` (default value: 0).

Reviewer: ```suggestion
- `maxburnamount` is a similar optional parameter for the `sendrawtransaction` RPC that helps prevent users from accidentally burning funds. Transactions with a data carrier `OP_RETURN` output amount that exceeds `maxburnamount` (default value: 0) are not submitted or broadcasted.
```

- This PR adds optional parameters `maxfeerate` and `maxburnamount` to the `submitpackage` RPC to perform similar sanity checks on package transactions.

- After decoding each package transaction in [`submitpackage`](https://github.com/bitcoin/bitcoin/blob/e69796c79c0aa202087a13ba62d9fbcc1c8754d4/src/rpc/mempool.cpp#L814), this PR adds a check that ensures that if a transaction output is unspendable and the output amount exceeds the `maxburnamount` value, the `submitpackage` function will abort and return immediately.

- The [`ProcessNewPackage`](https://github.com/bitcoin/bitcoin/blob/e69796c79c0aa202087a13ba62d9fbcc1c8754d4/src/validation.h#L286) function validates package transactions against consensus and policy rules. When invoked with `test_accept` set to `false`, all successfully validated transactions are added to the mempool.

> - This PR modifies `ProcessNewPackage` to accept a new optional parameter [`max_sane_feerate`](https://github.com/bitcoin-core-review-club/bitcoin/blob/bf6fbea9794db5e8d0adce8ff6182574808cea54/src/validation.cpp#L1658). Whenever a package transaction's individual feerate exceeds the `max_sane_feerate` value, this and the remaining unvalidated transactions will not be accepted into the mempool. 

Reviewer: Again, it would be helpful to have notes about how max feerate is checked for each RPC. Though optional.

Reviewer: You can also link to our review club meeting on 19339 which goes into more detail.

Author: Added the notes in latest push thank you

- Package transactions submitted to the `submitpackage` RPC successfully are accepted to the node's mempool, but broadcasted to peers individually because [Package relay](https://bitcoinops.org/en/topics/package-relay/) is currently a work in progress. Successful submission does not mean the transactions will propagate throughout the network whenever the transactions need to be relayed as a package.

- `maxfeerate` is an optional parameter for the RPC's `sendrawtransaction` and `testmempoolaccept`. It prevents transactions from being added to mempool and broadcasted to peers if the transaction feerate exceeds `maxfeerate`, avoiding unintentional payment of a high fee rate.

- `maxburnamount` is an optional parameter of `sendrawtransaction` that prevents broadcasting transactions with a data carrier `OP_RETURN` output amount that exceeds certain value. The default value of `maxburnamount` is 0.

- The PR adds optional parameters `maxfeerate` and `maxburnamount` to `submitpackage` RPC to perform similar sanity checks on package transactions.

> - After decoding each package transaction in [`submitpackage`](https://github.com/bitcoin/bitcoin/blob/e69796c79c0aa202087a13ba62d9fbcc1c8754d4/src/rpc/mempool.cpp#L814).

Author: Thanks fixed

    - The PR adds a check that ensures if a transaction output is unspendable and the output amount exceeds the `maxburnamount` value, the `submitpackage` function will abort and return immediately.

- The [`ProcessNewPackage`](https://github.com/bitcoin/bitcoin/blob/e69796c79c0aa202087a13ba62d9fbcc1c8754d4/src/validation.h#L286) function validates package transactions against consensus and policy rules. When invoked with `test_accept` set to `false`, all successfully validated transactions are added to the mempool.

    - This PR modifies `ProcessNewPackage` to accept another optional parameter [`max_sane_feerate`](https://github.com/bitcoin-core-review-club/bitcoin/blob/bf6fbea9794db5e8d0adce8ff6182574808cea54/src/validation.cpp#L1658). 
    - Whenever package transaction's individual feerate exceeds the `max_sane_feerate` value, the transaction, and the rest of the package transactions yet to be validated will not be accepted into the mempool.

>     - After package processing all transactions that are not successfully accepted into the mempool are not broadcasted to peers.

Author: I rephrase the note

## Questions

1. Did you review the PR? [Concept ACK, approach ACK, tested ACK, or NACK](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#peer-review)? What was your review approach?

2. Why is it important to perform these sanity checks on submitted packages?

3. Are there other important sanity checks apart from `maxburnamount` and `maxfeerate` that should performed on packages before they are accepted to the mempool?

> 4. The options `maxburnamount` and `maxfeerate` can prevent a transaction from entering the mempool and being relayed. Can we consider these options as policy rules?

Author: I dont think its just a simple "yes" answer, yes its a truthy / falsy answer but I think its  can spark further discussion in to what policy rules are? Maybe I will rephrase to
> The options `maxburnamount` and `maxfeerate` can prevent a transaction from entering the mempool and being relayed. Can we consider these options as policy rules? Why yes or no?

5. The [bf6fbea97](https://github.com/bitcoin-core-review-club/bitcoin/commit/bf6fbea9794db5e8d0adce8ff6182574808cea54) commit message states: "This allows subpackage processing and is compatible with future package RBF work." How is this compatible with future RBF work?

6. Why do we validate `maxfeerate` against the modified feerate instead of the base fee rate? 

> 7. We validate `maxfeerate` against the modified feerate of individual package transactions, not package feerate, this can sometimes be incentive incompatible.

Author: It does not see my discussion with @instagibbs  here https://github.com/bitcoin/bitcoin/pull/28950#discussion_r1444454903

> 7. We validate `maxfeerate` against the modified feerate of individual package transactions, not package feerate, this can sometimes be incentive incompatible.

Reviewer 2: Sorry, I misread the question, thanks for clarifying!

components: ["rpc"]
host: ismaelsadeeq
status: upcoming
commit: bf6fbea9794db5e8d0adce8ff6182574808cea54
---

## Notes

> - Package transactions submitted to the [`submitpackage`](https://bitcoincore.org/en/doc/26.0.0/rpc/rawtransactions/submitpackage/) RPC successfully are accepted to the node's mempool, but broadcasted to peers individually because [Package relay](https://bitcoinops.org/en/topics/package-relay/) is currently a work in progress. Successful submission does not mean the transactions will propagate.

Reviewer: ```suggestion
- Package transactions submitted to the [`submitpackage`](https://bitcoincore.org/en/doc/26.0.0/rpc/rawtransactions/submitpackage/) RPC successfully are accepted to the node's mempool, but broadcasted to peers individually because [package relay](https://bitcoinops.org/en/topics/package-relay/) is currently a work in progress. Successful submission does not mean the transactions will propagate.
```

- `maxfeerate` is an optional parameter for the [`sendrawtransaction`](https://github.com/bitcoin-core-review-club/bitcoin/blob/bf6fbea9794db5e8d0adce8ff6182574808cea54/src/rpc/mempool.cpp#L35) and [`testmempoolaccept`](https://github.com/bitcoin-core-review-club/bitcoin/blob/bf6fbea9794db5e8d0adce8ff6182574808cea54/src/rpc/mempool.cpp#L104) RPCs. Transactions exceeding this feerate won't be added to mempool and broadcasted to peers, helping prevent unintentional overpayment.

- `maxburnamount` is a similar optional parameter for the `sendrawtransaction` RPC that helps prevent users from accidentally burning funds. Transactions with a data carrier `OP_RETURN` output amount that exceeds `maxburnamount` (default value: 0) are not submitted or broadcasted.

> - Upon decoding a transaction in `sendrawtransaction` and `testmempoolaccept` RPCs, `maxburnamount` sanity check is immediately performed.

Reviewer: It's not really a sanity check
```suggestion
- Upon decoding a transaction in `sendrawtransaction` and `testmempoolaccept` RPCs, `maxburnamount` is checked immediatley.
```

- `testmempoolaccept` performs `maxfeerate` sanity check after package processing. It ensures that the individual base fee of all the successfully validated transactions [does not exceed](https://github.com/bitcoin-core-review-club/bitcoin/blob/bf6fbea9794db5e8d0adce8ff6182574808cea54/src/rpc/mempool.cpp#L218) the fee of the transaction size at `maxfeerate` fee rate.

- This PR adds optional parameters `maxfeerate` and `maxburnamount` to the [`submitpackage`](https://github.com/bitcoin-core-review-club/bitcoin/blob/bf6fbea9794db5e8d0adce8ff6182574808cea54/src/rpc/mempool.cpp#L816) RPC, enabling similar sanity checks on package transactions.

> - After [decoding each package transaction in `submitpackage`](https://github.com/bitcoin/bitcoin/blob/e69796c79c0aa202087a13ba62d9fbcc1c8754d4/src/rpc/mempool.cpp#L814), the transaction outputs are subjected to a `maxburnamount` sanity check.

Reviewer: Isn't this a duplicate of the point a few lines up?

Author: That is explaining how it's performed for `testmempoolaccept` and `sendrawtransaction` RPC?

It's similar because it's basically the same approach

- The PR introduces the `max_sane_feerate` parameter to the [`ProcessNewPackage`](https://github.com/bitcoin-core-review-club/bitcoin/blob/bf6fbea9794db5e8d0adce8ff6182574808cea54/src/validation.h#L282) function. `submitpackage` now passes the value of `maxfeerate` as the `max_sane_feerate` when calling `ProcessNewPackage`.

> - `ProcessNewPackage` is updated to forward the `max_sane_feerate` to [`AcceptPackage`](https://github.com/bitcoin-core-review-club/bitcoin/blob/bf6fbea9794db5e8d0adce8ff6182574808cea54/src/validation.cpp#L1450), which performs the sanity check during [`Sub package`](https://github.com/bitcoin-core-review-club/bitcoin/blob/bf6fbea9794db5e8d0adce8ff6182574808cea54/src/validation.cpp#L1425) evaluation.

Reviewer: It seems odd to point to `AcceptSubPackage` instead of the line in which the max feerate is checked?

Author: t is checked in two different places: when the sub-package is a single transaction and when it has multiple transactions in `AcceptSingleTransaction` and `AcceptMultipleTransactions`.

We determine which one to call in `AcceptSubPackage` and perform the actual check within that scope. 
I thought this could easily be seen after going through the code within the `AcceptSubPackage` function?

Author: Of course I will be nice if you could help in making this clearer, you know better :D 
Thank you

    - Transactions with individual modified feerates exceeding `max_sane_feerate` will not accepted into the mempool, along with its descendants, as they will be missing input during their evaluation.

## Questions

1. Did you review the PR? [Concept ACK, approach ACK, tested ACK, or NACK](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#peer-review)? What was your review approach?

2. Why is it important to perform these sanity checks on submitted packages?

3. Are there other important sanity checks apart from `maxburnamount` and `maxfeerate` that should be performed on packages before they are accepted to the mempool?

> 4. The options `maxburnamount` and `maxfeerate` can prevent a transaction from entering the mempool and being relayed. Can we consider these options as policy rules? Explain why?

Reviewer: ```suggestion
4. The options `maxburnamount` and `maxfeerate` can prevent a transaction from entering the mempool and being relayed. Can we consider these options as policy rules? Why or why not?
```

5. The [bf6fbea97](https://github.com/bitcoin-core-review-club/bitcoin/commit/bf6fbea9794db5e8d0adce8ff6182574808cea54) commit message states: "This allows subpackage processing and is compatible with future package RBF work." What makes this compatible with future RBF work?

6. Why do we validate `maxfeerate` against the modified feerate instead of the base fee rate? 

> 7. We validate `maxfeerate` against the modified feerate of individual package transactions, not package feerate, this can sometimes be incentive incompatible.

Reviewer: What do you mean by incentive incompatible? In this scenario, a user is submitting their own transaction to their own node. The only consideration for incentives would be that the user doesn't want to waste their own money.

Author: In a CFPF, the sponsor tx individual fee rate might be greater than `maxfeerate`; however, as a package, it could be very much lower. If the parents can be accepted as a sub-package, they will be, and the child (sponsor transaction will be rejected), making the fee bump much lower than intended.

---
see https://github.com/bitcoin/bitcoin/pull/28950#discussion_r1444454903 discussion with an example.

Reviewer: I see what you mean, but using "incentive compatible" is confusing. There is nobody to be compatible with when you're just checking the feerate of your own transaction. Just saying "inaccurate" would make more sense.

    - How does this happen?
    - Why is it not a current concern? 
    - Under what conditions might it become problematic?

> 8. Why is `maxfeerate` not validated before package processing? 

Reviewer: Maybe: "Why can't `maxfeerate` be checked immediately after decoding like `maxburnamount` is?"

components: ["rpc"]
host: ismaelsadeeq
status: upcoming
commit: bf6fbea9794db5e8d0adce8ff6182574808cea54
---

## Notes
- The `submitpackage` RPC receives topologically sorted transactions, a child, and its unconfirmed parents. It validates them against consensus and policy rules, if they passes these checks, the transactions are added to the mempool.

> - Packages submitted to the `submitpackage` RPC are only added to the node's mempool and not relayed to peers. Package relay is currently a work in progress.

Author: Yes, fixed wording

> - Packages submitted to the `submitpackage` RPC are only added to the node's mempool and not relayed to peers. Package relay is currently a work in progress.

Author: Noticed `BroadcastTransaction` comment needs to be updated
Opened a PR for it  https://github.com/bitcoin/bitcoin/pull/29308

- `maxfeerate` is an optional parameter for the RPCs `sendrawtransaction` and `testmempoolaccept`. Users will use it to ensure that the transaction they are adding to their mempools and broadcasting to the network does not exceed some feerate, preventing accidental payment of an unreasonable transaction fee rate.

> - `maxburnamount` is also an optional parameter of `sendrawtransaction` that prevents broadcasting transactions with a data carrier `OP_RETURN` output that exceeds a certain value. The default value of `maxburnamount` is 0.

Reviewer 2: ```suggestion
- `maxburnamount` is an optional parameter of `sendrawtransaction` that prevents broadcasting transactions with a data carrier `OP_RETURN` output amount that exceeds a certain value. The default value of `maxburnamount` is 0.
```

Also: keeping sentence constructions identical between both parameters (this and the previous one) makes it easier for readers, e.g. : "... is an optional parameter **for** the ... **RPCs**". We like small diffs.

Author: Fixed

> - The PR attempts to add optional parameters `maxfeerate` and `maxburnamount` to the `submitpackage` RPC to perform sanity checks on packages submitted to the mempool.

Reviewer 2: ```suggestion
- This PR adds optional parameters `maxfeerate` and `maxburnamount` to the `submitpackage` RPC to perform sanity checks on packages submitted to the mempool.
```

Author: Fixed thanks

## Questions

1. Did you review the PR? [Concept ACK, approach ACK, tested ACK, or NACK](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#peer-review)? What was your review approach?

2. Why is it important to perform these sanity checks on submitted packages?

3. Are there other important sanity checks apart from `maxburnamount` and `maxfeerate` that should performed on packages before they are submitted?

4. The options `maxburnamount` and `maxfeerate` can prevent a transaction from entering the mempool and being relayed. Can we consider these options as policy rules?

> 5. In [bf6fbea97](https://github.com/bitcoin/bitcoin/pull/28950/commits/bf6fbea9794db5e8d0adce8ff6182574808cea54), the commit message states, "This allows subpackage processing and is compatible with future package RBF work." How is this possible?

Reviewer 2: ```suggestion
5. The [bf6fbea97](https://github.com/bitcoin/bitcoin/pull/28950/commits/bf6fbea9794db5e8d0adce8ff6182574808cea54) commit message states: "This allows subpackage processing and is compatible with future package RBF work." How is this possible?
```

Does your "this" in "How is this possible?" refer to the subpackage processing, or to the compatibility with future RBF work, or both? Clarification would be good.

Author: I clarified the question to be more specific

> 6. What's the difference between modified feerate and absolute fee rate?

Reviewer 2: Is this question relevant enough to the PR? Perhaps a better question (e.g. to add to another one) is to ask which kind of fee rate is used?

7. Why are we validating against the modified feerate instead of the absolute fee rate?

8. In [bf6fbea97](https://github.com/bitcoin/bitcoin/pull/28950/commits/bf6fbea9794db5e8d0adce8ff6182574808cea54), we are validating `maxfeerate` against the modified feerate of individual package transactions, not package feerate, this can sometimes be incentive incompatible. 
    - How does this happen ?
    - Why isn't it a current concern ? 
    - Under what conditions might it become problematic ?

> 

Author: Added some notes on implementation, and taken your questions suggestions

> 9. Is there another approach from [bf6fbea97](https://github.com/bitcoin/bitcoin/pull/28950/commits/bf6fbea9794db5e8d0adce8ff6182574808cea54) ? What are the tradeoffs?

Reviewer 2: In a single-commit PR, no need to keep linking to the same commit

Author: Fixed
